// upbit-api-node/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "subscribe", {
  enumerable: true,
  get: function get() {
    return _quotation.subscribe;
  }
});

_Object$defineProperty(exports, "getTicker", {
  enumerable: true,
  get: function get() {
    return _quotation.getTicker;
  }
});

_Object$defineProperty(exports, "getMinCandles", {
  enumerable: true,
  get: function get() {
    return _quotation.getMinCandles;
  }
});

_Object$defineProperty(exports, "getCandles", {
  enumerable: true,
  get: function get() {
    return _quotation.getCandles;
  }
});

_Object$defineProperty(exports, "getTick", {
  enumerable: true,
  get: function get() {
    return _quotation.getTick;
  }
});

_Object$defineProperty(exports, "getOrderbook", {
  enumerable: true,
  get: function get() {
    return _quotation.getOrderbook;
  }
});

_Object$defineProperty(exports, "getMarketList", {
  enumerable: true,
  get: function get() {
    return _quotation.getMarketList;
  }
});

_Object$defineProperty(exports, "Exchange", {
  enumerable: true,
  get: function get() {
    return _exchange.default;
  }
});

exports.default = void 0;

var _quotation = require("./quotation");

var _exchange = _interopRequireDefault(require("./exchange"));

const upbit = {
  subscribe: _quotation.subscribe,
  getTicker: _quotation.getTicker,
  getMinCandles: _quotation.getMinCandles,
  getCandles: _quotation.getCandles,
  getTick: _quotation.getTick,
  getOrderbook: _quotation.getOrderbook,
  getMarketList: _quotation.getMarketList,
  Exchange: _exchange.default
};
exports.default = upbit;
